hub/config/src/main/java/uk/gov/ida/hub/config/data/S3ConfigSource.java

```diff
@@ -1,81 +1,122 @@
 import com.fasterxml.jackson.databind.ObjectMapper;
-import com.fasterxml.jackson.databind.module.SimpleModule;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListenableFutureTask;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import uk.gov.ida.hub.config.ConfigConfiguration;
+import uk.gov.ida.hub.config.configuration.SelfServiceConfig;
 import uk.gov.ida.hub.config.domain.remoteconfig.RemoteConfigCollection;
+import uk.gov.ida.hub.config.domain.remoteconfig.RemoteConnectedServiceConfig;
+import uk.gov.ida.hub.config.domain.remoteconfig.RemoteMatchingServiceConfig;
+import uk.gov.ida.hub.config.domain.remoteconfig.RemoteServiceProviderConfig;
+import uk.gov.ida.hub.config.domain.remoteconfig.SelfServiceMetadata;
 
 import java.io.IOException;
 import java.io.InputStream;
-import java.text.DateFormat;
 import java.text.SimpleDateFormat;
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
 
 public class S3ConfigSource {
     private static final Logger LOG = LoggerFactory.getLogger(S3ConfigSource.class);
-    private ConfigConfiguration configConfiguration;
+    public static final Map<String, RemoteConnectedServiceConfig> EMPTY_CONNECTED_SERVICE_CONFIG_MAP = Collections.emptyMap();
+    public static final Map<String, RemoteMatchingServiceConfig> EMPTY_MATCHING_SERVICE_CONFIG_MAP = Collections.emptyMap();
+    public static final List<RemoteServiceProviderConfig> EMPTY_SERVICE_PROVIDER_CONFIG_LIST = Collections.emptyList();
+    public static final RemoteConfigCollection EMPTY_COLLECTION = new RemoteConfigCollection(null, null, EMPTY_CONNECTED_SERVICE_CONFIG_MAP, EMPTY_MATCHING_SERVICE_CONFIG_MAP, EMPTY_SERVICE_PROVIDER_CONFIG_LIST);
+    private SelfServiceConfig selfServiceConfig;
     private AmazonS3 s3Client;
+    private LoadingCache<String, RemoteConfigCollection> cache;
+    private ObjectMapper objectMapper;
 
     public S3ConfigSource(ConfigConfiguration configConfiguration, AmazonS3 s3Client) {
-        this.configConfiguration = configConfiguration;
-        this.s3Client = s3Client;
+        this.selfServiceConfig = configConfiguration.getSelfService();
+        if (selfServiceConfig != null && selfServiceConfig.isEnabled()){
+            this.s3Client = s3Client;
+            this.objectMapper = new ObjectMapper();
+            objectMapper.setDateFormat(new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss"));
+
+            CacheLoader<String, RemoteConfigCollection> cacheLoader = new S3ConfigCacheLoader();
+            long expiry = selfServiceConfig.getCacheExpiryInSeconds();
+            TimeUnit unit = TimeUnit.SECONDS;
```

Comment:
Dropwizard has support for doing these type conversion using the Duration type. See [here](https://github.com/alphagov/verify-hub/blob/master/hub/policy/src/main/java/uk/gov/ida/hub/policy/configuration/PolicyConfiguration.java#L31) for an example

Reply:
Oh, nice. I've used that.